### PR TITLE
 Fix #920: Preset deprecated search engines remain sticky/set.

### LINKED ISSUE #920: Default search engines that were removed are being reverted to region defaults
### Description: 

Upgrading `1.7.4 (19.02.01.22)` --> `1.8 (19.02.20.22)` will always revert the search engine to the regions default (either Google/Qwant) even though the user has already picked another default from the preexisting list.

### Steps to Reproduce 

  1. download `1.7.4 (19.02.01.22)` from the App Store and open Brave
  2. change the default search engine to `Amazon` and the PB default to `Yahoo`
  3. upgrade to `1.8 (19.02.20.22)` using the `ipa` file that was provided
  4. relaunch brave and you'll notice that the default search engine is either Google/Qwant

**Actual result:** 

Currently reverting the default search engine to either Google/Qwant every time.

**Expected result:**

If a user has selected another search engine from the preexisting list, we should honour that choice.

**Reproduces how often:** [Easily reproduced, Intermittent Issue]

100% reproducible using the above STR.

**Brave Version:** 

Upgrading from 1.7.4 (19.02.01.22)` --> `1.8 (19.02.20.22)`

**Device details:** 

Used `iPad Mini 4 running iOS 12.1.4`

### PR DESCRIPTION
**Pull request by jhreis**
Fix #920 

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)

// Unit tests not needed as this is not an ongoing supported item
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

File: Client.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -399,8 +399,8 @@
 		55A747171DC46FC400CE1B57 /* HomePageUITest.swift in Sources */ = {isa = PBXBuildFile; fileRef = 55A747161DC46FC400CE1B57 /* HomePageUITest.swift */; };
 		595E0EDB21CAD97C00813D49 /* CookieControl.js in Resources */ = {isa = PBXBuildFile; fileRef = 595E0EDA21CAD97C00813D49 /* CookieControl.js */; };
 		595E0EE121CAEF5B00813D49 /* FileManagerExtension.swift in Sources */ = {isa = PBXBuildFile; fileRef = 595E0EE021CAEF5B00813D49 /* FileManagerExtension.swift */; };
-		59871C1921E4D83900316590 /* block-cookies.json in Resources */ = {isa = PBXBuildFile; fileRef = 59871C1821E4D83900316590 /* block-cookies.json */; };
 		5965CA222215D4A0000B9EC1 /* CustomHeaderDataTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 5965CA212215D4A0000B9EC1 /* CustomHeaderDataTests.swift */; };
+		59871C1921E4D83900316590 /* block-cookies.json in Resources */ = {isa = PBXBuildFile; fileRef = 59871C1821E4D83900316590 /* block-cookies.json */; };
```
Comment (iccub) [inline Client.xcodeproj/project.pbxproj:403]: Why was this touched? I guess just a xcode thing

File: Client/Frontend/Browser/Search/DefaultSearchPrefs.swift [hunk 2]
```diff
@@ -59,13 +59,35 @@ class DefaultSearchPrefs {
...
         // Overrides for specfic regions.
         if let overrides = regionOverrides[region].dictionary {
             usersEngineList = usersEngineList.map({ overrides[$0]?.string ?? $0 })
         }
-        return usersEngineList
+        
+        return usersEngineList.unique { $0 == $1 }
+    }
+    
+    private func filePaths(fromShortNames shortNames: [String]) -> [String]? {
```
Comment (iccub) [inline Client/Frontend/Browser/Search/DefaultSearchPrefs.swift]: Is this method name correct? It returns search engine names, not paths imo

File: Client/Frontend/Browser/Search/DefaultSearchPrefs.swift [hunk 3]
```diff
@@ -59,13 +59,35 @@ class DefaultSearchPrefs {
...
+        do {
+            let filePath = URL(fileURLWithPath: path)
+            let jsonData = try Data(contentsOf: filePath)
+            let mappings = try JSONDecoder().decode([String: String].self, from: jsonData)
+            return shortNames.compactMap { mappings[$0] }
+        } catch {
+            // Fail silently, nothing needed for recovery here really
+        }
+        
+        return nil
```
Comment (iccub) [inline Client/Frontend/Browser/Search/DefaultSearchPrefs.swift]: I would remove optional from method return, and return `[]` here so we don't need to unwrap it on line 65

File: Client/Frontend/Browser/Search/DefaultSearchPrefs.swift [hunk 4]
```diff
@@ -59,13 +59,35 @@ class DefaultSearchPrefs {
...
+            return nil
+        }
+        
+        do {
+            let filePath = URL(fileURLWithPath: path)
+            let jsonData = try Data(contentsOf: filePath)
+            let mappings = try JSONDecoder().decode([String: String].self, from: jsonData)
+            return shortNames.compactMap { mappings[$0] }
+        } catch {
+            // Fail silently, nothing needed for recovery here really
```
Comment (iccub) [inline Client/Frontend/Browser/Search/DefaultSearchPrefs.swift:87]: Add `log.error` it doesn't need to be that silent :)

File: Client/Frontend/Browser/Search/SearchEngines.swift [hunk 5]
```diff
@@ -261,7 +261,8 @@ class SearchEngines {
     /// Get all known search engines, possibly as ordered by the user.
     fileprivate func getOrderedEngines() -> [OpenSearchEngine] {
         let locale = Locale(identifier: Locale.preferredLanguages.first ?? Locale.current.identifier)
-        let unorderedEngines = customEngines + SearchEngines.getUnorderedBundledEnginesFor(locale: locale)
+        let prefs = [Preferences.Search.defaultEngineName, Preferences.Search.defaultPrivateEngineName].compactMap { $0.value }
```
Comment (iccub) [inline Client/Frontend/Browser/Search/SearchEngines.swift]: This name is not clear imo, I had to take a moment to understand what it is.
Maybe rename it, and method parameters you added to something like `selectedSearchEngines`?

**Review by iccub — CHANGES_REQUESTED** [review]
Logic looks good, I have just few suggestions about better naming and nitpicks

File: Client.xcodeproj/project.pbxproj (hunk 1 above)
Comment (jhreis) [reply Client.xcodeproj/project.pbxproj:403]: Yeah, not sure why. I did add a file to the bundle, so probably just adjusted bundles files? 🤷‍♂️

File: Client/Frontend/Browser/Search/DefaultSearchPrefs.swift (hunk 4 above)
Comment (jhreis) [reply Client/Frontend/Browser/Search/DefaultSearchPrefs.swift:87]: Just don't want to place unnecessary error logs, really do not care if this fails for some unknown or random reason.

**Review by iccub — APPROVED** [review]

**Comment by iccub** [comment]
Approving, but unit tests still need to be updated to make travis pass

**Pull request merged**